### PR TITLE
Adding retry to DPS device registration operations

### PR DESCRIPTION
**Pull request by timtay-microsoft**
This PR adds a passive retry mechanism to all device registration operations. If a provisioning device client receives a transient exception (429 and above) during registration status polling, the sdk will log that it happened, and then continue to poll for updates until a final status is reached.

### CODE REVIEW

File: provisioning/transport/http/src/Generated/RuntimeRegistration.cs [hunk 1]
```diff
@@ -159,13 +159,22 @@ public RuntimeRegistration(DeviceProvisioningServiceRuntimeClient client)
             HttpStatusCode _statusCode = _httpResponse.StatusCode;
             cancellationToken.ThrowIfCancellationRequested();
             string _responseContent = null;
+
+            if ((int)_statusCode == 429 || (int)_statusCode >= 500)
+            {
+                var ex = new ProvisioningTransportException("Received status code " + _statusCode, null, true);
```
Comment (rajeevmv) [inline provisioning/transport/http/src/Generated/RuntimeRegistration.cs]: >ProvisioningTransportException [](start = 29, length = 30)

Will this cause a retry? i.e., is this a transient exception?

File: provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs [hunk 2]
```diff
@@ -309,7 +322,25 @@ private async Task VerifyPublishPacketTopicAsync(IChannelHandlerContext context,
...
+                    if (statusCode == HttpStatusCode.BadRequest)
+                    {
+                        await FailWithExceptionAsync(
+                         context,
+                         new ProvisioningTransportException(
+                             jsonData,
+                             null,
+                             false)).ConfigureAwait(false);
+                    }
+                    else if ((int)statusCode == 429)
```
Comment (rajeevmv) [inline provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs]: >(int)statusCode == 429) [](start = 29, length = 23)

looks like this condition is covered in line 347 below.

File: provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs [hunk 3]
```diff
@@ -309,7 +322,25 @@ private async Task VerifyPublishPacketTopicAsync(IChannelHandlerContext context,
             {
                 if (Enum.TryParse(match.Groups[1].Value, out HttpStatusCode statusCode))
                 {
-                    if (statusCode >= HttpStatusCode.BadRequest)
+                    if (statusCode == HttpStatusCode.BadRequest)
```
Comment (rajeevmv) [inline provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs]: >BadRequest [](start = 53, length = 10)

even this condition appears to be covered by line 347

File: provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs (hunk 2 above)
Comment (timtay-microsoft) [reply provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs]: As per your note [here](https://github.com/Azure/azure-iot-sdk-csharp/pull/750#discussion_r249187971) aren't 429 and 400 special cases where we cannot parse into a ProvisioningErrorDetails object? So don't we need these special cases where we do not parse into a json object?

Comment (rajeevmv) [reply provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs]: Sorry, I was answering the question of whether the response would contain the status of 'Assigning'.  Anush is correct and you should receive an object of type 'ProvisioningServiceErrorDetails' which you are deserializing on line 345.

---
In reply to: [255747413](https://github.com/Azure/azure-iot-sdk-csharp/pull/791#discussion_r255747413) [](ancestors = 255747413)

Comment (timtay-microsoft) [reply provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs]: Ah okay, I'll leave the error parsing as it was, then.

File: provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs (hunk 3 above)
Comment (timtay-microsoft) [reply provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs]: I'll revert this to maintain the current error parsing then

**Comment by timtay-microsoft** [comment]
@rajeevmv I have updated the PR to no longer treat 400/429 as special cases. Can you take one last look today?

**Comment by timtay-microsoft** [comment]
//AzurePipelines run all

**Comment by timtay-microsoft** [comment]
Closing this since the feature was abandoned for now

**Pull request closed without merging**